EIPS/eip-6807.md

```diff
@@ -0,0 +1,116 @@
+  
+  ```javascript
+    // Returns the array of this structure
+    {
+      contractAddress: "0x...abc",
+      name: "Your contract name",
+      description?: "your contract description",
+      code?: "Link to sol file of this contract"
+    }[]
+  ```
+  
+  Further, DAD must deploy a DRC that has the following structure:
+  
+  ```solidity
+    interface IDAppRegistry {
+      
+      function isMyContract(address _address) external view returns (bool);
+    }
+  ```
+  
+  We define the Registry contract as:
+  
+  ```solidity
+    contract DomainContractRegistry {
+      struct RegistryInfo {
+        address dappRegistry;
+        address admin;
+      }
+      
+      mapping(string => RegistryInfo) public registryMap;
+  
+      function setDappRegistry(string memory _domain, address _dappRegistry) external {
+        // check if a domain already has a dapp registry mapped
+        // if yes, check if owner is same. if so, allow the change
+        // if no, recordTransition(_domain, _dappRegistry);
+        
+        IDappRegistry dappRegistry = IDappRegistry(_dappRegistry);
+        // Use chainlink to the call `{{domain}}/contracts.json`. 
+        // Check for each contract listed in contracts.json by calling 
+        // dappRegistry.isMyContract(_address)
+        
+        // If all addresses match, update registryMap
+      }
+      
+      // In case of a domain transfer, register that there is potential change in registry mapping
+      // and a new owner may be attempting to update registry
+      // A cool-off period is applied on the domain marking a potential transfer in ownership
+      // Cool-off period can be 7 days
+      function recordTransition(string memory _domain, address _dappRegistry) internal
+       
+    }
+  ```
+  
+  DAD must register their domain in this registry to validate domain ownership. 
+
+## Rationale
+
+Wallets need official contract addresses of a domain to warn users if the domain is sending transaction to a different address. This could have been solved by allowing DADs to provide a standard url (e.g. /contracts.json) to wallets. However, in the event of a DNS attack, even this information can be tampered. By deploying a registry contract on chain, the DAD is able to have a second source of their official contracts that they can set when they have full control of their domain. If registry's information gets tampered, wallets can always use standard url to cross-check. This system ensures the attackers has to get access to both admin private keys and domain control to do the attack which is highly difficult relatively. This is analogous to 2FA. 
+  
+<!-- More details - TBD -->
+
+## Backwards Compatibility
+
+No backward compatibility issues found.
+
+## Test Cases
+
+<!-- TODO -->
+
+<!--
+
+## Reference Implementation
+
+-->
```

Comment:
We don't allow external links. If you would like to include a reference implementation, please do so in the `../assets/eip-6807` folder with a CC0-1.0 license.